tests/unit/recommenders/evaluation/test_python_time.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+import pandas as pd
+import pytest
+from unittest.mock import Mock
+from sklearn.preprocessing import minmax_scale
+
+from recommenders.utils.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL,
+    DEFAULT_RATING_COL,
+    DEFAULT_PREDICTION_COL,
+    SEED,
+)
+from recommenders.evaluation.python_evaluation import (
+    merge_rating_true_pred,
+    merge_ranking_true_pred,
+    rmse,
+    mae,
+    rsquared,
+    exp_var,
+    get_top_k_items,
+    precision_at_k,
+    recall_at_k,
+    ndcg_at_k,
+    map_at_k,
+    auc,
+    logloss,
+)
+import random
+from recommenders.utils.timer import Timer
+
+random.seed(SEED)
+DATA_USER_NUM = 10000
+DATA_ITEM_NUM = DATA_USER_NUM * 2
+DATA_SAMPLE_NUM = DATA_USER_NUM * 1000
+DATA_RATING_MAX = 5
+
+# fmt: off
+DF_RATING_TRUE = pd.DataFrame(
+    {
+        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_ITEM_COL: random.choices(range(0, DATA_ITEM_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_RATING_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
+    }
+)
+
+
+DF_RATING_PRED = pd.DataFrame(
+    {
+        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_ITEM_COL: random.choices(range(0, DATA_ITEM_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_PREDICTION_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
+        # DEFAULT_RATING_COL: random.choices(range(1, RATING_MAX+1), k=SAMPLE_NUM),
+    }
+)
+
+
+DF_RATING_NOHIT = pd.DataFrame(
+    {
+        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
+        DEFAULT_ITEM_COL: [DATA_ITEM_NUM] * DATA_SAMPLE_NUM,
+        DEFAULT_PREDICTION_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
+    }
+)
+
+
+@pytest.fixture
+def rating_true():
+    return DF_RATING_TRUE.copy()
```

Comment:
The typical way to do what you want to do is you create a fixture and create the dataframe inside, without the copy and without the global variables.
```
@pytest.fixture
def rating_pred():
    return pd.DataFrame(
    {
        DEFAULT_USER_COL: random.choices(range(0, DATA_USER_NUM), k=DATA_SAMPLE_NUM),
        DEFAULT_ITEM_COL: random.choices(range(0, DATA_ITEM_NUM), k=DATA_SAMPLE_NUM),
        DEFAULT_PREDICTION_COL: random.choices(range(1, DATA_RATING_MAX+1), k=DATA_SAMPLE_NUM),
        # DEFAULT_RATING_COL: random.choices(range(1, RATING_MAX+1), k=SAMPLE_NUM),
    }
)
```
There are several scope levels for fixtures: https://github.com/miguelgfierro/pybase/blob/main/test/pytest_scope.py the default one is function, which means that the df is generated everytime the fixture is called inside a test function. If you are not going to change any of the DF, you can add the scope module, which is that the DF is computed and added to memory once, and then used for all the tests inside this file

Reply:
Updated